### DISCUSSION TITLE
Why xsd:anyURI literals instead of real URLs?

### DISCUSSION THREAD

**Discussion by VladimirAlexiev** [edmcouncil/fibo#1367] [General]
- me (from #983 item 17): Please explain why do you use owl:DatatypeProperties with ^^xsd:anyURI instead of real URLs (i.e. owl:ObjectProperties without range).
- @ElisaKendall If you use OWL reasoning, there are certain declarations / uses that disallow a URI in the range. So for annotations we've used anyURI.

Elisa, can you please explain these declarations / uses / restrictions?
- I've wondered for years why people use xsd:anyURI literals instead of real URLs, and haven't found an explanation. 
- It's worth explaining it in https://github.com/edmcouncil/fibo/blob/master/ONTOLOGY_GUIDE.md

I analyzed instances of anyURI in 2020Q3-prod like this:
```
 find . -iname '*.ttl'|xargs cat|riot -syntax ttl -out nt |sort > all.nt
 grep anyURI all.nt |cut -d " " -f2 |sort|uniq -c|sort -rn > anyURI-props.txt
```

A lot are annotation props but not all. Star indicates maybe a change is needed:
```
*  1990 <https://spec.edmcouncil.org/fibo/ontology/FND/Places/VirtualPlaces/hasWebsite>
*   571 <http://www.omg.org/techprocess/ab/SpecificationMetadata/dependsOn>
    479 <https://spec.edmcouncil.org/fibo/ontology/FND/Utilities/AnnotationVocabulary/adaptedFrom>
*   130 <http://purl.org/dc/terms/license>
*   122 <http://www.omg.org/techprocess/ab/SpecificationMetadata/contentLanguage>
     41 <https://spec.edmcouncil.org/fibo/ontology/FND/Utilities/AnnotationVocabulary/definitionOrigin>
      4 <http://www.w3.org/2000/01/rdf-schema#seeAlso>
      2 <http://www.w3.org/2000/01/rdf-schema#range>
      1 <https://spec.edmcouncil.org/fibo/ontology/FND/Utilities/AnnotationVocabulary/explanatoryNote>
      1 <http://www.w3.org/2004/02/skos/core#example>
      1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first>
      1 <http://www.omg.org/techprocess/ab/SpecificationMetadata/relatedSpecification>
      1 <http://www.omg.org/techprocess/ab/SpecificationMetadata/directSource>
?     1 <http://purl.org/dc/terms/source>
```

I checked range and the list:
```
 grep anyURI all.nt |grep -E "#range|#first"
```
The following props have range xsd:anyURI:
- <https://spec.edmcouncil.org/fibo/ontology/FND/Places/VirtualPlaces/hasURL> 
- <https://spec.edmcouncil.org/fibo/ontology/FND/Places/VirtualPlaces/hasWebsite>
- <https://spec.edmcouncil.org/fibo/ontology/FND/Utilities/Analytics/QualifiedMeasure> (string or anyURI, see #933)

**Comment by VladimirAlexiev** [comment]
Why changes are needed:

A. (from #983 item 18) `sm:dependsOn` should be an object property since its target is an object (a owl:Ontology). It's bad style to represent the same resource as anyURI literal and as IRI:
```ttl
	sm:dependsOn
"https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/"^^xsd:anyURI 
	owl:imports
<https://spec.edmcouncil.org/fibo/ontology/master/latest/FND/Accounting/CurrencyAmount/> 
```

B. contentLanguage uses both of these forms, which is bad:
```
"http://www.w3.org/standards/techs/owl#w3c_all"^^<http://www.w3.org/2001/XMLSchema#anyURI>
<http://www.w3.org/standards/techs/owl#w3c_all>
```

C. DCT specifies `license range dct:LicenseDocument` (an object) and `source a rdf:Property` (which admits literal or object)

D. `hasWebsite` and `hasURL`: If the criterion for using anyURI is "does the target have a type" or "does the target have some statements", that's a very brittle criterion and contradicts OWA.
Eg if I add a type and some statements about a website URL (eg using schema.org), does that mean `hasWebsite` should become an object property?

**Reply by ElisaKendall** [reply]
@VladimirAlexiev As you hopefully have noticed, we've eliminated all references to the SpecificationMetadata in favor of a combination of the Commons Annotation Vocabulary extended with a few things in FIBO, but far fewer than we used to have. I think we've addressed most of these issues you've raised since this was first reported, FYI.

**Comment by rivettp** [comment]
These are all good points except I don't understand D - it seems that these are declared and used as datatype properties with the anyURI datatype. I don't think it makes sense to try to represent sites as resources for FIBO. And if they were separately modeled as resources I don't see that being a problem either.

**Comment by VladimirAlexiev** [comment]
@rivettp what you describe in D means that FIBO rejects OWA.

-  It's ok for FIBO to decide that website info is out of scope of FIBO, 
- but it's not ok to select a representation such that **nobody** can add info about those websites

**Comment by mdebellis** [comment]
I just started looking at FIBO but I've worked and built many OWL ontologies, so what I'm about to say may not be the way FIBO looks at it but from the standard of what people typically do with OWL ontologies the first point is that an IRI (that is the term we're supposed to use now, it's the same as a URI except it supports additional character sets) is not the same thing as a URL. A URL is a sub-type of an IRI. All URLs are IRIs but not all IRIs are URLs. I don't know the specific distinction in format but in terms of use cases, a URL is usually a web page, so fairly large grained. An IRI can be for a Class, Property, etc. In an OWL ontology there is the IRI for the ontology like https://ontology and each entity in the ontology has its own IRI like https://ontology/My_Class and https://ontology/my_property. People tend to use xsd:anyURI for annotation or data properties where we are pointing at things (usually not other ontology entities) outside the ontology, in fact much of the time those data or annotation properties do point to HTML pages and you can use their URL as a value because URLs are also IRIs. I think that is why people use xsd:anyURI because that will accept URLs just fine but if you need to use a non-URL IRI you still have that option. 

In my experience people never put values for URLs into object properties. In OWL that wouldn't work but that gets to my question:

From what I've seen (and this is just a very brief look so I may have picked an unusual file for my first look) FIBO looks like it is using RDF and RDFS not OWL. E.g., in the file I was looking at, I think it was for business processes there were lots of definitions for rdfs:Class rather than owl:Class. An RDFS class is not the same as an OWL class. E.g., you can't run a reasoner on it. Is this true for all of FIBO or does FIBO also have ontologies with owl classes as well?

**Reply by ElisaKendall** [reply]
@mdebellis Hi Michael, Welcome to FIBO! I think we have a good understanding of URIs, IRIs, and URLs. In FIBO, I think the only remaining places where we use xsd:anyURI,  are in annotations. Some of those, such as for dct:license, are being replaced with the full license text so many of those remaining will go away.

With respect to the use of rdfs:Class, where did you find that? There are none in the GitHub repo for FIBO that I'm aware of. There may be some in the version of FIBO posted to the EDM Council web site at spec.edmcouncil.org, in the version that has been translated to SKOS, but not in any of the main ontologies.

I'd be happy to give you a walk-through at some point - would love to share more about Commons as well as the current version of FIBO, how we are using patterns from Commons and adding more over time, etc. Feel free to reach out to me directly at ekendall@thematix.com, and I'd be delighted to set something up. I'm in Los Altos, in the SF Bay Area, with respect to time zones for calls, and typically afternoons are better for me, but let me know what works for you and I'll look forward to chatting with you.
